SQL data sources: Convert to data frames

ying-jeanne opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Convert SQL data sources to data frames.

TODO:

- [x] Fix annotations
- [x] Fix frontend tests

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #30362

**Special notes for your reviewer**:
Delivering of data from the backend currently doesn't work, i.e. you get empty graphs.

Commit:
convert dataframe

wbrowne committed:
support datetime + timestamp

wbrowne committed:
fix up test

wbrowne committed:
resolve conflicts

wbrowne committed:
Merge branch 'master' into 30362

Commit:
add update for mssql

Commit:
fix postgres converter except for time column to milisecond

marefr commented on pkg/tsdb/mysql/mysql.go:
```diff
@@ -150,3 +160,154 @@ func (t *mysqlQueryResultTransformer) TransformQueryError(err error) error {
+		},
+	},
+	{
+		Name:           "handle DATETIME",
```
Feels really weird that the `sqlutil.FrameFromRows` cannot handle this. Seems like this is what causes a string to be returned, but not sure?
https://github.com/grafana/grafana-plugin-sdk-go/blob/b638fd747d60136049a428cff146ddeea1239a27/data/sqlutil/sql.go#L119

Would really like to understand if there's potentially something missing in the sqlutil package to support more types out of the box.

kylebrandt replied:
I think it is because `database/sql` defers types to the driver. So what you get from a scan will depend on the database, and the driver used.

So in this case `*string` is really a synonm for []byte (since our DF doesn't have []byte).

I am curious what the types of these `InputKindScan` reflect.Struct's are. If they are things like [`database/sql` `NullTime`](https://golang.org/pkg/database/sql/#NullTime) then supporting those in `sqlutil.FrameFromRows` probably makes sense.

ying-jeanne replied:
point noted, we could do a clean up to sdk later on as a second step.

kylebrandt commented on pkg/tsdb/mysql/mysql.go:
```diff
@@ -150,3 +160,154 @@ func (t *mysqlQueryResultTransformer) TransformQueryError(err error) error {
+		Replacer: &sqlutil.StringFieldReplacer{
+			OutputFieldType: data.FieldTypeNullableFloat64,
+			ReplaceFunc: func(in *string) (interface{}, error) {
+				spew.Dump(in)
```
be sure to remove the spew lines eventually :P

kylebrandt commented on pkg/tsdb/mysql/mysql_test.go:
```diff
@@ -139,40 +139,41 @@ func TestMySQL(t *testing.T) {
-				So(column[28], ShouldEqual, nil)
-				So(column[29], ShouldEqual, "")
-				So(column[30], ShouldEqual, nil)
+				frames, _ := queryResult.Dataframes.Decoded()
```
The data package has `data.FrameTestCompareOptions()` meant to be used with `go-cmp`.  

This way you can construct the frame you are expecting as a whole thing.

For an example (in a table test in this case) see https://github.com/grafana/grafana/blob/05fa9d6ad9dd2eff5bae1e5b82806b2f27a8f7a3/pkg/tsdb/azuremonitor/azuremonitor-datasource_test.go#L444-L446

Commit:
add time series interpolate

Commit:
seprate trim and resample and reput unittests

ying-jeanne commented on pkg/tsdb/mssql/mssql_test.go:
```diff
@@ -1011,11 +1018,13 @@ func TestMSSQL(t *testing.T) {
 
 				// Should be in milliseconds
-				So(columns[0].(int64), ShouldEqual, dt.Unix()*1000)
+				// So(columns[0].(int64), ShouldEqual, dt.Unix()*1000)
```
those lines are commented out, currently i am not able to get the int64 when running the test, I suppose it is because i am not longer calling ConvertSqlValueColumnToFloat() and ConvertSqlTimeColumnToEpochMs(). Need to understand why we need to convert and how to do the equivalent in dataframe

Commit:
Convert time column from other format to timet`

Commit:
intermedia

Commit:
convert all columns except for time and metric to float64

Commit:
for god sake postgres unittest all passed

Commit:
fix mssql test

Commit:
fix all mssql test

Commit:
fix sqleng unittest

Commit:
fix all mysql test

Commit:
Merge remote-tracking branch 'origin/master' into 30362-wb

Commit:
Merge remote-tracking branch 'origin/master' into 30362-wb

Commit:
fix conflict

Commit:
fix go-lint

aknuds1 commented:
Is this still a draft @ying-jeanne ?

aknuds1 commented:
Handing this over to Erik, since pkg/tsdb is the domain of cloud data sources.

marefr commented on pkg/tsdb/postgres/postgres_test.go:
```diff
@@ -151,8 +152,8 @@ func TestGenerateConnectionString(t *testing.T) {
 	// change to true to run the PostgreSQL tests
-	runPostgresTests := false
-	// runPostgresTests := true
+	// runPostgresTests := false
+	runPostgresTests := true
```
```suggestion
	runPostgresTests := false
```

marefr commented on pkg/tsdb/mysql/mysql_test.go:
```diff
@@ -32,741 +33,746 @@ import (
 func TestMySQL(t *testing.T) {
 	// change to true to run the MySQL tests
-	runMySQLTests := false
+	runMySQLTests := true
```
```suggestion
	runMySQLTests := false
```

marefr reviewed (commented):
Added a couple suggestions that should hopefully fix the drone/CI issues

ying-jeanne committed:
Update pkg/tsdb/postgres/postgres_test.go

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

ying-jeanne committed:
Update pkg/tsdb/mysql/mysql_test.go

Co-authored-by: Marcus Efraimsson <marcus.efraimsson@gmail.com>

Commit:
Merge remote-tracking branch 'origin' into 30362-wb

Commit:
remove useless functions + fix time with zone

Commit:
fix go-lint

Commit:
Merge branch '30362-wb' of https://github.com/grafana/grafana into 30362-wb

Commit:
the draft frontend changes

sunker commented:
@ying-jeanne was there a reason not to use [sqlds](https://github.com/grafana/sqlds) for the core sql data sources? Is it only because it's private?

Commit:
revert macro test change

marefr commented on pkg/tsdb/sqleng/trim.go:
```diff
@@ -0,0 +1,53 @@
+)
+
+// trim trims rows that are outside the qm.TimeRange
+func trim(f *data.Frame, qm DataQueryModel) error {
```
We didn't have this before, why is it needed now? Something missing in the re-sample code that you have to take heights for in this trim function?

marefr commented on pkg/tsdb/sqleng/trim.go:
```diff
@@ -0,0 +1,53 @@
+			return fmt.Errorf("time point is nil")
+		}
+
+		if t.(time.Time).Before(qm.TimeRange.From) {
```
Did we do this before? I can't find it. Think we should return all data that the user/query returns

marefr reviewed (commented):
Think this looks pretty promising :+1: Added some comments/suggestions based on our discussions

Commit:
fix template variable dropdown list + dataframe return as table

ying-jeanne commented on pkg/tsdb/sqleng/sql_engine.go:
```diff
@@ -41,10 +37,10 @@ type SQLMacroEngine interface {
 	// TransformQueryError transforms a query error.
 	TransformQueryError(err error) error
+
+	GetConverterList() []sqlutil.StringConverter
```
I might missed something, and in function dataquery i didn't see how I can access to DataPluginConfiguration

ying-jeanne commented:
> @ying-jeanne was there a reason not to use [sqlds](https://github.com/grafana/sqlds) for the core sql data sources? Is it only because it's private?

there are 3 parts that currently sqlds is not supporting as mentioned in the weekly, but we need in SQL data sources
1. when return type is time series, user can select any column type as time. when the column is numeric, we convert it into time.Time in the backend.
2. when the return type is time series, we convert all value columns into float64 to keep good precision and also to support easily timegroup filling float64 when the point is missing.
3. all these 3 SQL datasources doesn't have built-in function to insert missing points, that is what resample function is doing. that is currently not supported in sqlds.

Commit:
clean up frontend unused functions

Commit:
fix error in backend

Commit:
remove the check on result meta data

Commit:
deactivate postgres unnittest

Commit:
remove unnittest change

Commit:
fix converters in mysql

Commit:
Merge remote-tracking branch 'origin/master' into 30362-wb

Commit:
fix unittest frontend

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Clean up

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
TSDB: Simplify

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
TSDB: Simplify

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
TSDB: Simplify

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Refactor

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Fix tests

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Simplify

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Cleanup

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Fix error handling

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

Commit:
fix frontend mssql/postgres

Commit:
fix frontend

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

alexanderzobnin committed:
Access control: Build navigation links with access control (#33024)

* Build nav links with access control

* Break up getNavTree (reduce cyclomatic complexity)

* Fix tests

* Use only ActionUsersRead permissions

* Remove unused permissions definitions

* Chore: remove unused fallbacks

* Fix linter error

domasx2 committed:
Alerting: misc fixes (#33070)

domasx2 committed:
Alerting: receivers page + template list (#33112)

dependabot[bot] committed:
Bump github.com/uber/jaeger-client-go from 2.25.0+incompatible to 2.26.0+incompatible (#33106)

* Bump github.com/uber/jaeger-client-go

Bumps [github.com/uber/jaeger-client-go](https://github.com/uber/jaeger-client-go) from 2.25.0+incompatible to 2.26.0+incompatible.
- [Release notes](https://github.com/uber/jaeger-client-go/releases)
- [Changelog](https://github.com/jaegertracing/jaeger-client-go/blob/master/CHANGELOG.md)
- [Commits](https://github.com/uber/jaeger-client-go/compare/v2.25.0...v2.26.0)

Signed-off-by: dependabot[bot] <support@github.com>

* Fix go.sum

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

hugohaggmark committed:
Tests: fixes issue with fetch

hugohaggmark committed:
ConfirmModal: Migrates to React with new theme (#33107)

* ConfirmModal: Migrates to React

* Refactor: migrates to v2 styles

* Chore: updates after PR comments

Elfo404 committed:
Elasticsearch:  Add documentation for supported Elasticsearch query transformations (#33072)

* Elasticsearch:  Add documentation for supported Elasticsearch query transformation

* Update docs/sources/datasources/elasticsearch.md

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

* Remove beta notives for ES logs

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

aocenas committed:
Plugins: Support es5 plugins extending LanguageProvider (#33127)

thisisobate committed:
ValuePicker: refactors story from knobs to control (#33122)

* ValuePicker: refactors story from knobs to control

* fixes wrong type issue

FZambia committed:
Live: Telegraf input modifiers (#32982)

ryantxu committed:
PanelEdit: only update when config changes (#33133)

FZambia committed:
bump jaeger-client-go to v2.27.0 (#33134)

marefr committed:
Auth: Don't clear auth token cookie when lookup token fails (#32999)

If LookupToken fails we don't clear the auth token cookie.

Ref #15316

oddlittlebird committed:
Docs: Update pie-chart-panel.md (#33132)

owen-d committed:
Inhouse alerting api (#33129)

* init

* autogens AM route

* POST dashboards/db spec

* POST alert-notifications spec

* fix description

* re inits vendor, updates grafana to master

* go mod updates

* alerting routes

* renames to receivers

* prometheus endpoints

* align config endpoint with cortex, include templates

* Change grafana receiver type

* Update receivers.go

* rename struct to stop swagger thrashing

* add rules API

* index html

* standalone swagger ui html page

* Update README.md

* Expose GrafanaManagedAlert properties

* Some fixes

- /api/v1/rules/{Namespace} should return a map
- update ExtendedUpsertAlertDefinitionCommand properties

* am alerts routes

* rename prom swagger section for clarity, remove example endpoints

* Add missing json and yaml tags

* folder perms

* make folders POST again

* fix grafana receiver type

* rename fodler->namespace for perms

* make ruler json again

* PR fixes

* silences

* fix Ok -> Ack

* Add id to POST /api/v1/silences (#9)

Signed-off-by: Ganesh Vernekar <cs15btech11018@iith.ac.in>

* Add POST /api/v1/alerts (#10)

Signed-off-by: Ganesh Vernekar <cs15btech11018@iith.ac.in>

* fix silences

* Add testing endpoints

* removes grpc replace directives

* [wip] starts validation

* pkg cleanup

* go mod tidy

* ignores vendor dir

* Change response type for Cortex/Loki alerts

* receiver unmarshaling tests

* ability to split routes between AM & Grafana

* api marshaling & validation

* begins work on routing lib

* [hack] ignores embedded field in generation

* path specific datasource for alerting

* align endpoint names with cloud

* single route per Alerting config

* removes unused routing pkg

* regens spec

* adds datasource param to ruler/prom route paths

* Modifications for supporting migration

* Apply suggestions from code review

* hack for cleaning circular refs in swagger definition

* generates files

* minor fixes for prom endpoints

* decorate prom apis with required: true where applicable

* Revert "generates files"

This reverts commit ef7e97558477d79bcad416e043b04dbd04a2c8f7.

* removes server autogen

* Update imported structs from ngalert

* Fix listing rules response

* Update github.com/prometheus/common dependency

* Update get silence response

* Update get silences response

* adds ruler validation & backend switching

* Fix GET /alertmanager/{DatasourceId}/config/api/v1/alerts response

* Distinct gettable and postable grafana receivers

* Remove permissions routes

* Latest JSON specs

* Fix testing routes

* inline yaml annotation on apirulenode

* yaml test & yamlv3 + comments

* Fix yaml annotations for embedded type

* Rename DatasourceId path parameter

* Implement Backend.String()

* backend zero value is a real backend

* exports DiscoveryBase

* Fix GO initialisms

* Silences: Use PostableSilence as the base struct for creating silences

* Use type alias instead of struct embedding

* More fixes to alertmanager silencing routes

* post and spec JSONs

* Split rule config to postable/gettable

* Fix empty POST /silences payload

Recreating the generated JSON specs fixes the issue
without further modifications

* better yaml unmarshaling for nested yaml docs in cortex-am configs

* regens spec

* re-adds config.receivers

* omitempty to align with prometheus API behavior

* Prefix routes with /api

* Update Alertmanager models

* Make adjustments to follow the Alertmanager API

* ruler: add for and annotations to grafana alert (#45)

* Modify testing API routes

* Fix grafana rule for field type

* Move PostableUserConfig validation to this library

* Fix PostableUserConfig YAML encoding/decoding

* Use common fields for grafana and lotex rules

* Add namespace id in GettableGrafanaRule

* Apply suggestions from code review

* fixup

* more changes

* Apply suggestions from code review

* aligns structure pre merge

* fix new imports & tests

* updates tooling readme

* goimports

* lint

* more linting!!

* revive lint

Co-authored-by: Sofia Papagiannaki <papagian@gmail.com>
Co-authored-by: Domas <domasx2@gmail.com>
Co-authored-by: Sofia Papagiannaki <papagian@users.noreply.github.com>
Co-authored-by: Ganesh Vernekar <15064823+codesome@users.noreply.github.com>
Co-authored-by: gotjosh <josue@grafana.com>
Co-authored-by: David Parrott <stomp.box.yo@gmail.com>
Co-authored-by: Kyle Brandt <kyle@grafana.com>

dependabot[bot] committed:
Bump github.com/centrifugal/centrifuge from 0.16.0 to 0.17.0 (#33117)

kminehart committed:
add cache config tab to datasource settings (#33036)

ryantxu committed:
GraphNG: flip logic so we can explicitly say the field types to keep (#33140)

aknuds1 committed:
Fix frontend

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Fix frontend

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Fix frontend

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Fix frontend

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=32257) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=32257) before we can accept your contribution.<br/>**0** out of **4** committers have signed the CLA.<br/><br/>:x: wbrowne<br/>:x: ying-jeanne<br/>:x: aknuds1<br/>:x: hugohaggmark<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=32257) it.</sub>

marefr committed:
MSSQL: Dataframe client side migration

marefr commented:
Done with MSSQL client side migration I think. Everything seems to work. Just have to try running the mssql_tests devenv docker block and check related gdev dashboard.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Fix frontend

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

marefr committed:
MSSQL: Fix tests

marefr committed:
MSSQL: Fix annotations and metricfindquery

marefr committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

marefr commented:
Merging with master and fixing merge conflicts

aknuds1 committed:
Merge branch '30362-wb' of github.com:grafana/grafana into 30362-wb

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Fix frontend

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

marefr committed:
Fix mysql data source type errors

marefr committed:
Merge remote-tracking branch 'origin/master' into arve/30362

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Undo unintended changes

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

marefr committed:
Fix mssql tests

marefr committed:
Fix mysql and tests

marefr committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

marefr committed:
Fix mysql datetime converters

marefr committed:
Fix lint problem

marefr committed:
Fix postgres and tests

marefr committed:
Use data frames response in mysql tests

marefr committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

marefr committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

ying-jeanne commented:
LGTM, the backend test is better to use go-cmp to compare entire dataframe, but it is probably better handle in another pr. tested with 3 datasources, test seems fine. @ericuldall I would let eric take a look at the frontend code.

kylebrandt commented on pkg/tsdb/sqleng/resample.go:
```diff
@@ -0,0 +1,128 @@
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+)
+
+func getRowFillValues(f *data.Frame, tsSchema data.TimeSeriesSchema, currentTime time.Time,
```
I wonder if we are replicating behavior already done in the sdk, since LongToWide takes fillmissing 

https://github.com/grafana/grafana-plugin-sdk-go/blob/867dddf55bd1438063435f15a1638f1bf6c2d605/data/time_series.go#L206

https://github.com/grafana/grafana-plugin-sdk-go/blob/master/data/time_series.go#L290

kylebrandt replied:
Oh I think I see, this some additional post processing for some reason :-) :+1:

marefr commented:
@ying-jeanne I suppose you meant to ping @sunker, right?

marefr committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

marefr committed:
Cleanup

marefr committed:
Add missing executeQuery recovery

marefr committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

marefr committed:
fix

aknuds1 commented on pkg/tsdb/sqleng/sql_engine.go:
```diff
@@ -237,86 +173,172 @@ func (e *dataPlugin) DataQuery(ctx context.Context, dsInfo *models.DataSource,
+	defer func() {
+		if r := recover(); r != nil {
+			e.log.Error("executeQuery panic", "error", r, "stack", log.Stack(1))
+			if theErr, ok := r.(error); ok {
+				queryResult.Error = theErr
+				ch <- queryResult
+			} else if theErrString, ok := r.(string); ok {
+				queryResult.Error = fmt.Errorf(theErrString)
+				ch <- queryResult
+			}
```
Shouldn't we deal with the case where `r` is neither `error` nor `string`? In that case, the panic will be swallowed.

marefr replied:
Was a bit unsure about that. What are you suggesting? `ch <- queryResult` on all panics, but setting `queryResult.Error` if string or error?

aknuds1 replied:
Since we don't have any clear way to convert other types than string to `error`, I would suggest we either re-panic or assign a generic error to `queryResult.Error`.

marefr replied:
I'll go with a generic error since re-panic would cause the same issue that I tried to solve with this recovery - not crashing the go-routine. We had that problem with mssql library trying to use a non-supported data type for example if you remember.

aknuds1 commented on pkg/tsdb/sqleng/sql_engine.go:
```diff
@@ -326,413 +348,533 @@ func newProcessCfg(query plugins.DataSubQuery, queryContext plugins.DataQuery, r
+)
+
+type dataQueryModel struct {
+	InterpolatedQuery string // property non set until after Interpolate()
```
```suggestion
	InterpolatedQuery string // property not set until after Interpolate()
```

aknuds1 commented on pkg/tsdb/sqleng/sql_engine.go:
```diff
@@ -326,413 +348,533 @@ func newProcessCfg(query plugins.DataSubQuery, queryContext plugins.DataQuery, r
+	InterpolatedQuery string // property non set until after Interpolate()
+	Format            dataQueryFormat
+	TimeRange         backend.TimeRange
+	FillMissing       *data.FillMissing // property non set until after Interpolate()
```
```suggestion
	FillMissing       *data.FillMissing // property not set until after Interpolate()
```

aknuds1 commented on public/app/plugins/datasource/postgres/datasource.ts:
```diff
@@ -71,40 +72,24 @@ export class PostgresDatasource {
-    if (queries.length === 0) {
-      return of({ data: [] });
+  filterQuery(query: PostgresQuery): boolean {
+    if (query.hide) {
+      return false;
     }
+    return true;
```
Nit, but simpler code:
```suggestion
    return !query.hide;
```

aknuds1 commented on public/app/plugins/datasource/mysql/response_parser.ts:
```diff
@@ -120,47 +85,40 @@ export default class ResponseParser {
+    const timeField = frame.fields.find((f) => f.name === 'time' || f.name === 'time_sec');
+
+    if (!timeField) {
+      return Promise.reject({ message: 'Missing mandatory time column (with time column alias) in annotation query.' });
```
`Promise.reject` should be the same as throwing an exception in an `async` function, you should also use `Error` objects for exceptions/rejections:
```suggestion
      throw new Error('Missing mandatory time column (with time column alias) in annotation query');
```

aknuds1 commented on public/app/plugins/datasource/mysql/response_parser.ts:
```diff
@@ -120,47 +85,40 @@ export default class ResponseParser {
-        message: 'Missing mandatory time column (with time_sec column alias) in annotation query.',
-      };
+    if (frame.fields.find((f) => f.name === 'title')) {
+      return Promise.reject({
+        message: 'The title column for annotations is deprecated, now only a column named text is returned',
+      });
```
```suggestion
      throw new Error('The title column for annotations is deprecated, now only a column named text is returned');
```

aknuds1 commented on public/app/plugins/datasource/postgres/response_parser.ts:
```diff
@@ -102,45 +88,34 @@ export default class ResponseParser {
-        message: 'Missing mandatory time column in annotation query.',
-      });
+    if (!timeField) {
+      return Promise.reject({ message: 'Missing mandatory time column (with time column alias) in annotation query.' });
```
```suggestion
      throw new Error('Missing mandatory time column (with time column alias) in annotation query');
```

aknuds1 reviewed (changes requested):
Generally looks good to me! I spotted some typos though, and I think that in `async` functions you should throw exceptions instead of returning rejected promises, right? Rejected promises and async exceptions should functionally be the same, but it's better to consistently stick to the new style IMO.

aknuds1 commented:
Build fails BTW, in frontend tests.

marefr committed:
Fix mssql query ctrl

marefr committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

marefr committed:
Apply suggestions from code review

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

marefr committed:
return generic error in panic recovery

marefr committed:
Merge remote-tracking branch 'origin/master' into 30362-wb

marefr committed:
mysql: fix annotation editor

marefr committed:
postgres: export plugin in non-legacy way

sunker commented on pkg/tsdb/sqleng/resample.go:
```diff
@@ -0,0 +1,128 @@
+	}
+	return vals
+}
+
```
It's still not crystal clear to me what this and the `getRowFillValues` function does. Could a comment that describes that be added?

marefr replied:
I pushed some updates. Not sure they make it easier to understand, but will ask Ying to review it.

sunker commented on public/app/plugins/datasource/mssql/response_parser.ts:
```diff
@@ -1,73 +1,41 @@
 
-    return this.transformToSimpleList(rows);
+    return Array.from(new Set(values.map((v) => v.text))).map((text) => {
+      return { text, value: values.find((v) => v.text === text)?.value } as MetricFindValue;
```
Don't think there's a need for this cast?

marefr replied:
Good catch. Pushed fix

sunker reviewed (commented):
Tested with Postgres and it works fine. 

Just found one nit.

marefr committed:
fixes after review

marefr committed:
code comments

marefr reviewed (approved):
Tested all three of the data sources the best I could and everything seems to work as before. LGTM

marefr committed:
Update pkg/tsdb/mssql/mssql_test.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 reviewed (approved):
LGTM!

hugohaggmark commented:
This PR causes empty results in variables with just integers #35391 
https://github.com/grafana/grafana/blob/6d78f92816ce7453a9b446a4ab469a6649d649c2/public/app/plugins/datasource/postgres/response_parser.ts#L25
